content/posts/batching-requests.md

```diff
@@ -0,0 +1,186 @@
+
+We will create a route with the path `/speaker` that returns both the sessions and topics of a particular speaker using the `public-api` plugin. You can do this using the [ApisixRoute](https://apisix.apache.org/docs/ingress-controller/concepts/apisix_route/) CRD:
+
+```yaml {title="speaker-route.yaml"}
+apiVersion: apisix.apache.org/v2
+kind: ApisixRoute
+metadata:
+  name: speaker-route
+spec:
+  http:
+    - name: speaker
+      match:
+        paths:
+          - /speaker
+      plugins:
+        - name: public-api
+          enable: true
+          config:
+            uri: /apisix/batch-requests
+      backends:
+        - serviceName: apisix-admin
+          servicePort: 9180
+```
+
+> **Note**: The ApisixRoute CRD always requires you to configure an upstream to a route. To work around this, you can use the Admin API as the upstream, as shown above, in scenarios where you don't have/need an upstream.
+
+The above route extracts the individual requests needed to be made by APISIX. To fulfill these requests using the upstream, we have to create another route:
+
+```yaml {title="speaker-topics-sessions-route.yaml"}
+apiVersion: apisix.apache.org/v2
+kind: ApisixRoute
+metadata:
+  name: speaker-topics-sessions-route
+spec:
+  http:
+    - name: speaker-topics-sessions
+      match:
+        hosts:
+          - conferenceapi.azurewebsites.net
+        paths:
+          - /*
+        exprs:
+          - subject:
+              scope: Path
+              name: topics
+            op: RegexMatch
+            value: "^/speaker/[^/]*/topics$"
+          - subject:
+              scope: Path
+              name: sessions
+            op: RegexMatch
+            value: "^/speaker/[^/]*/sessions$"
+        methods:
+          - "GET"
+      upstreams:
+        - name: conference-api-upstream
+```
+
+This may seem complicated, but it is simple. We are matching requests using a regular expression in the path. So regardless of the value of `{speakerId}` in `/speaker/{speakerId}/sessions` and `/speaker/{speakerId}/topics`, we will match the request and forward it to the upstream.
+
+> **Note**: If you are configuring APISIX through the Admin API, this is much simpler and reduces to `"uris": ["/speaker/*/topics","/speaker/*/sessions"]`. However, we must use regular expressions since the ApisixRoute CRD does not support [URI patterns](https://github.com/apache/apisix-ingress-controller/blob/c111c12272f58189de785feac256e6ce1193c172/samples/deploy/crd/v1/ApisixRoute.yaml#L107C35-L107C35) with `*` in the middle.
+
+## Test the Routes
+
+To test the route, we can send a request to the exposed `/speaker` endpoint:
+
+```shell
+curl -i http://127.0.0.1:50404/speaker -X POST -d \
+'{
+  "pipeline": [
+    {
+      "method": "GET",
+      "path": "/speaker/1/topics"
+    },
+    {
+      "method": "GET",
+      "path": "/speaker/1/sessions"
+    }
+  ]
+}'
+```
+
+APISIX will make both of these requests and return the aggregated response as shown below:
```

Comment:
>APISIX will make both of these requests

APISIX will fork this single HTTP request into two requests for each endpoint. The forking happens closer to the upstreams, and so saves network traffic compared to if the client had send two separate requests.

(Or something like this, re-explain the benefit)